smartconfig/entry.py

```diff
@@ -0,0 +1,138 @@
+from typing import Any, Dict, NoReturn, Optional, Tuple
+
+from smartconfig import _registry
+from smartconfig._registry import configuration_for_module, global_configuration, lookup_global_configuration
+from smartconfig.exceptions import ConfigurationKeyError, InvalidOperation, PathConflict
+
+
+class _ConfigEntryMeta(type):
+    """
+    Metaclass used to define special ConfigEntry behaviors.
+
+    Note: Using this metaclass outside of the library is currently not supported.
+
+    The lookup of attributes can be customized by subclassing this metaclass and changing `_get_attribute_path`.
+    """
+
+    def __getattribute__(cls, name: str) -> Optional[Any]:
+        """
+        Look up the attribute through the configuration system.
+
+        If the attribute name starts with `_`, normal lookup is done, otherwise _registry.global_configuration is used.
+
+        Args:
+            name: Attribute to lookup
+
+        Raises:
+            ConfigurationKeyError: The attribute doesn't exist.
+        """
+        # Use the normal lookup for attribute starting with `_`
+        if name.startswith('_'):
+            return super().__getattribute__(name)
+
+        path, attribute = cls._get_attribute_path(name)
+
+        return lookup_global_configuration(path, attribute)
+
+    def __new__(cls, name: str, bases: Tuple[type, ...], dict_: Dict[str, Any], path: Optional[str] = None) -> type:
+        """
+        Add the `__defined_attributes` and `__path_override` attributes to the new entry.
+
+        Args:
+            path: Custom path used for this entry instead of the module name.
+        """
+        names = chain(dict_.keys(), dict_.get('__annotations__', {}).keys())
+        dict_[f"{cls.__name__}__defined_attributes"] = {name for name in names if not name.startswith('_')}
+
+        dict_[f"{cls.__name__}__path_override"] = path
+
+        return super().__new__(cls, name, bases, dict_)
+
+    def _register_entry(cls) -> None:
+        """Set the `__path` attribute and register the entry."""
+        cls.__path = cls.__path_override or cls.__module__
+        if cls.__path in _registry.configuration_for_module:
+            raise PathConflict(f"An entry at {cls.__path!r} already exists.")  # TODO: Add an FAQ link.
+
+        configuration = {
+            key: value for key, value in cls.__dict__.items() if not key.startswith('_')
+        }
+
+        if cls.__path not in global_configuration:
+            global_configuration[cls.__path] = configuration
+        # We already have some overrides for this path.
+        else:
+            for key, value in configuration.items():
+                path, key = cls._get_attribute_path(key)
+
+                if path not in global_configuration:
+                    global_configuration[path] = {}
+
+                # We only write values that aren't already defined.
+                if key not in global_configuration[path]:
+                    global_configuration[path][key] = value
```

Comment:
With the current setup, they don’t get removed, but I don’t think it is much of an issue since adding the default isn’t a destructive operation, and reloading the file would result in the same global configuration.

Reply:
If there's a !REF to an attribute of a class that gets unloaded, then the reference will incorrectly resolve to the stale value rather than raising an exception.

Reply:
We can solve this by removing attributes when the class is unloaded only if the value is still the default (so we don't override values coming from the configuration). We can hook that up on `__del__` but we are relying on the garbage collector. What do you think about this solution?

Reply:
I think relying on GC is fine. However, how will you know if the value is still the default? If the config happens to have the same value as the default, then that value should still be accessible if the class unloads since it's not relying on the class for the value.

Reply:
Resolved by 20f3e2b05982ee0afb98c87d353cdf09c2751abd